src/tbb/threading_control.cpp

```diff
@@ -0,0 +1,397 @@
+        if (thr_control != nullptr) {
+            return;
+        }
+
+        // POSSIBLE MEMORY LEAK THERE!!!!!!!
+        thr_control = new (cache_aligned_allocate(sizeof(threading_control))) threading_control(1, 1);
+        thr_control->my_pimpl = d1::make_cache_aligned_unique<threading_control_impl>(thr_control);
+
+        __TBB_InitOnce::add_ref();
+
+        if (global_control_active_value_unsafe(global_control::scheduler_handle)) {
+            ++thr_control->my_public_ref_count;
+            ++thr_control->my_ref_count;
+        }
+
+        g_threading_control = thr_control;
+        }).on_completion([] { global_control_unlock(); });
+
+    return thr_control;
+}
+
+void threading_control::destroy () {
+    d1::cache_aligned_deleter deleter;
+    deleter(this);
+    __TBB_InitOnce::remove_ref();
+}
+
+void threading_control::wait_last_reference(global_mutex_type::scoped_lock& lock) {
+    while (my_public_ref_count.load(std::memory_order_relaxed) == 1 && my_ref_count.load(std::memory_order_relaxed) > 1) {
+        lock.release();
+        // To guarantee that request_close_connection() is called by the last external thread, we need to wait till all
+        // references are released. Re-read my_public_ref_count to limit waiting if new external threads are created.
+        // Theoretically, new private references to the threading control can be added during waiting making it potentially
+        // endless.
+        // TODO: revise why the weak scheduler needs threading control's pointer and try to remove this wait.
+        // Note that the threading control should know about its schedulers for cancellation/exception/priority propagation,
+        // see e.g. task_group_context::cancel_group_execution()
+        while (my_public_ref_count.load(std::memory_order_acquire) == 1 && my_ref_count.load(std::memory_order_acquire) > 1) {
+            yield();
+        }
+        lock.acquire(g_threading_control_mutex);
+    }
+}
+
+bool threading_control::release(bool is_public, bool blocking_terminate) {
+    bool do_release = false;
+    {
+        global_mutex_type::scoped_lock lock(g_threading_control_mutex);
+        if (blocking_terminate) {
+            __TBB_ASSERT(is_public, "Only an object with a public reference can request the blocking terminate");
+            wait_last_reference(lock);
+        }
+        do_release = remove_ref(is_public);
+    }
+
+    if (do_release) {
+        __TBB_ASSERT(!my_public_ref_count.load(std::memory_order_relaxed), "No public references remain if we remove the threading control.");
+        // inform RML that blocking termination is required
+        my_pimpl->release(blocking_terminate);
+        return blocking_terminate;
+    }
+    return false;
+}
+
+threading_control::threading_control(unsigned public_ref, unsigned ref) : my_public_ref_count(public_ref), my_ref_count(ref)
+{}
+
+threading_control* threading_control::register_public_reference() {
```

Comment:
Seems strange that the `register_public_reference` returns something.

Reply:
So order should be -> if obj exist - yes -> inc ref - no > create it -> inc ref -> return obj
Do you propose create two different function? One for creation and another just getter? They should follow each other everywhere